### ISSUE TITLE
SLPVectorizer's PHICompare doesn't provide a strict weak ordering

### ISSUE THREAD

**Issue by dwblaikie** [llvm/llvm-project#79321]
https://github.com/llvm/llvm-project/blame/fc364e26845ce5529caf9f88abcc5a5531d1f59f/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L4138

Discovered via an internal bug where we don't have a good reproduction for this yet, but eyeballing the comparator it appears it could be buggy so I figured I'd start with this bug.

The issue was flagged by libc++ comparator checking 

There are several places in the comparator where it tests "if either element is missing this property, then they are not-less-than (ie: they are equal)" but assuming all states are valid/there's no external restrictions on the state of these objects, that definition can lead to an invalid/non-strict-weak comparator. 

Take the first case of this in the comparator: `!hasOneUse` - if either `Value` has more than one use, they are considered not-less-than.

But if you had `A`, `B`, and `C` - where `B` `!hasOneUse`, but `A` and `C` `hasOneUse` - then you end up with a situation where `A /< B` and `B /< C`, but `A < C`, potentially.

The issue comes up several times in this function - pretty much all the `return false` (including the trailing one, which catches both the dyn_cast check pairs, so counts for two instances of the problem) have this issue - so 7 instances by the looks of it.

The correct way to do these checks would be to choose an ordering for these cases - in the `hasOneUse` example, for instance - all the `!hasOneUse` might be the earlier elements, so it'd look like changing this code:

```
if (!V1->hasOneUse() || !V2->hasOneUse())
  return false;
```
To this:
```
if (!V1->hasOneUse() < !V2->hasOneUse())
  return true;
if (!V1->hasOneUse() > !V2->hasOneUse())
  return false;
...
```

I don't think it can be done in fewer steps - I think it requires at least two comparisons per pair like this. (that could be wrapped up in some utility to make it easier, like:
```
std::optional<bool> ThreeWayBoolCompare(bool A, bool B) {
  if (A < B)
    return true;
  if (B < A)
    return false;
  return std::nullopt;
}
...
if (std::optional<bool> Comp = ThreeWayBoolCompare(!V1->hasOneUse(), !V2->hasOneUse()))
  return *Comp;
```

**Comment by dwblaikie** [comment]
Hmm - actually it's the /other/ `PHICompare` in thta file that we're seeing the failure from... 
https://github.com/llvm/llvm-project/blame/fc364e26845ce5529caf9f88abcc5a5531d1f59f/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L15926
Looking into that one. At first glance it looks a bit more strict-weak...

**Comment by dwblaikie** [comment]
Yeah, at least some of these things don't seem to be strict-weak, same as above, like:
```
if (isa<Instruction>(Opcodes1[I]))
  return true;
if (isa<Instruction>(Opcodes2[I]))
  return false;
```
If they're both instructions, then `A < B` and `B < A` would both be true, which is a contradiction.

**Comment by dwblaikie** [comment]
The specific stack where this failed is:
```
#6			__check_strict_weak_ordering_sorted<llvm::Value **, llvm::function_ref<bool (llvm::Value *, llvm::Value *)> >	include/c++/v1/__debug_utils/strict_weak_ordering_check.h:58
#7	(inlined)	__stable_sort_impl<std::__u::_ClassicAlgPolicy, llvm::Value **, llvm::function_ref<bool (llvm::Value *, llvm::Value *)> >	include/c++/v1/__algorithm/stable_sort.h:252
#8	(inlined)	stable_sort<llvm::Value **, llvm::function_ref<bool (llvm::Value *, llvm::Value *)> >	include/c++/v1/__algorithm/stable_sort.h:258
#9	(inlined)	stable_sort<llvm::SmallVectorImpl<llvm::Value *> &, llvm::function_ref<bool (llvm::Value *, llvm::Value *)> >	llvm/include/llvm/ADT/STLExtras.h:1981
#10	tryToVectorizeSequence<llvm::Value>	third_party/llvm/llvm-project/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:15683
#11	vectorizeChainsInBlock	llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:16064
```

Using libc++ with `_LIBCPP_DEBUG_STRICT_WEAK_ORDERING_CHECK` enabled.

**Comment by alexey-bataev** [comment]
Tried to fix it in 36e4a7ecca8a79fe8960a1e2afa78fbbd643dc7f, please check if it fixes the issue.

**Comment by dwblaikie** [comment]
My internal user has checked and doesn't have the issue anymore (but there's local file changes too, etc).

So the acute problem is addressed, though I am still a bit uncertain about the code/patch. I'll see if I can walk through my thinking here:
```
   for (int I = 0, E = Opcodes1.size(); I < E; ++I) {
      // Undefs are compatible with any other value.
      if (isa<UndefValue>(Opcodes1[I]) || isa<UndefValue>(Opcodes2[I])) {
        if (isa<UndefValue>(Opcodes1[I]) && isa<UndefValue>(Opcodes2[I]))
          continue;
        if (isa<Instruction>(Opcodes1[I]))
          return true;
        if (isa<Instruction>(Opcodes2[I]))
          return false;
        if (isa<Constant>(Opcodes1[I]) && !isa<UndefValue>(Opcodes1[I]))
          return true;
        if (isa<Constant>(Opcodes2[I]) && !isa<UndefValue>(Opcodes2[I]))
          return false;
        if (isa<UndefValue>(Opcodes1[I]) && !isa<UndefValue>(Opcodes2[I]))
          return false;
        if (!isa<UndefValue>(Opcodes1[I]) && isa<UndefValue>(Opcodes2[I]))
          return true;
        continue;
      }
      ...
```

So if the two things are both undef, they're considered equal as far as this property is concerned.
But if the LHS is an Instruction, it's less than the undef. (instruction is less than undef)
If the RHS is an Instruction, it's greater than the undef. (undef is not less than instruction, aka instruction is less than or equal to undef)
OK, so that puts instructions on the left/"less" end of the sort.

Then the constant cases... 
LHS is constant nonundef on the LHS is less than... anything?
RHS constant nonundef is not less than... anything?
That doesn't seem to make sense to me - if both sides are constant not-undef, shouldn't that return false in both cases?

**Comment by alexey-bataev** [comment]
> Then the constant cases...
LHS is constant nonundef on the LHS is less than... anything?
RHS constant nonundef is not less than... anything?
That doesn't seem to make sense to me - if both sides are constant not-undef, shouldn't that return false in both cases?

You forgot about outer if condition. If LHS is const nonundef, RHS must be undef and vice versa. I'll add assertions.

**Comment by alexey-bataev** [comment]
Added some asserts (+ simplified the code) here 35f45926eb12b2612f23012ab1e80e3be86dbb2d

**Comment by dwblaikie** [comment]
Hmm, fair enough. 

I wonder if there'd be a way to write it more clearly - like does that first block need to check the instruction/constant cases, or could it look more like?

```
if both undef
  continue
if rhs undef
  return true (undefs are greater than all non-undefs)
if lhs undef
  return false (undef is not less than anything nonundef)
then on to all the instruction and constant handling
```
Also the rest of the handling seems to mix up the ordering of instructions and constants (it does instructions, then constants, then instructions relative to constants, then back to constants) - might be simpler if those were grouped together
```
if both instructions
  // test ordering of instructions
if lhs instruction
  return true (instructions are less than constants
if rhs instruction
  return false (constants are not less than instructions)
  
assert both constants
// test ordering of constant ids
```
Then it's more uniform/easier to read
* check the undef ordering
* if neither undef
* check the instruction ordering
* if neither instruction
* check the constant ordering.

**Comment by alexey-bataev** [comment]
Originally the code looked pretty similar way. But the problem is that while sorting the values, we need to group them very carefully. Undefs must follow the instructions, constants must follow undefs and instructions, all other values do not matter too much. Instructions themselves must be grouped by their opcodes and/or their alternative opcodes to handle as much vectorization opportunities as possible. Undefs, instructions with the same/alternative opcodes, constants and/or undefs can be vectorized.
I tried to land simplified comparator, by it leads to a very-hard-to-detect transitive comparison issues. Prefer to have it in a more explicit way, actually much easier to understand the issue and fix it.

**Comment by dwblaikie** [comment]
So the order of instructions < undefs < constants doesn't seem to describe the code in its current state, for instance:
```
        if (isa<Constant>(Opcodes2[I]) && !isa<UndefValue>(Opcodes2[I])) {
          assert(isa<UndefValue>(Opcodes1[I]) && "Expected 1st undef value");
          return false;
        }
```

This condition causes `undef < constant` to evaluate to false, suggesting that the constant is greater than the undef... I'm not sure if the rest of the function is consistent with this definition.

It's all these special cases that I think makes it hard to follow the logic of the code to ensure it does provide the desired ordering.

I /think/ the ordering this provides is instructions < constants < undef... 

And it looks like maybe non-instruction non-constants go before undef?
```
if (!isa<UndefValue>(Opcodes2[I])) {
  assert(isa<UndefValue>(Opcodes1[I]) && "Expected 1st undef value");
  return false;
}
```

So I prototyped an alternative implementation - and asserted that it produced the same results as the current implementation - and came up with this, which I think is easier to understand/clearer:

```
            // Instructions come first.
            auto *I1 = dyn_cast<Instruction>(Opcodes1[I]);
            auto *I2 = dyn_cast<Instruction>(Opcodes2[I]);
            if (I1 && I2) {
              DomTreeNodeBase<BasicBlock> *NodeI1 = DT->getNode(I1->getParent());
              DomTreeNodeBase<BasicBlock> *NodeI2 = DT->getNode(I2->getParent());
              if (!NodeI1)
                return NodeI2 != nullptr ? Less : Greater;
              if (!NodeI2)
                return Greater;
              assert((NodeI1 == NodeI2) ==
                         (NodeI1->getDFSNumIn() == NodeI2->getDFSNumIn()) &&
                     "Different nodes should have different DFS numbers");
              if (NodeI1 != NodeI2)
                return NodeI1->getDFSNumIn() < NodeI2->getDFSNumIn() ? Less
                                                                     : Greater;
              InstructionsState S = getSameOpcode({I1, I2}, *TLI);
              if (S.getOpcode() && !S.isAltShuffle())
                return Equal;
              return I1->getOpcode() < I2->getOpcode() ? Less : Greater;
            }
            if (I1)
              return Less;
            if (I2)
              return Greater;
          }
          {
            // Non-undef constants come next.
            bool C1 = isa<Constant>(Opcodes1[I]) && !isa<UndefValue>(Opcodes1[I]);
            bool C2 = isa<Constant>(Opcodes2[I]) && !isa<UndefValue>(Opcodes2[I]);
            if (C1 && C2)
              return Equal;
            if (C1)
              return Less;
            if (C2)
              return Greater;
          }
          bool U1 = isa<UndefValue>(Opcodes1[I]);
          bool U2 = isa<UndefValue>(Opcodes2[I]);
          {
            // Non-constant non-instructions come next.
            if (!U1 && !U2) {
              auto ValID1 = Opcodes1[I]->getValueID();
              auto ValID2 = Opcodes2[I]->getValueID();
              if (ValID1 == ValID2)
                return Equal;
              if (ValID1 < ValID2)
                return Less;
              if (ValID1 > ValID2)
                return Greater;
            }
            if (!U1)
              return Less;
            if (!U2)
              return Greater;
          }
          // Undefs come last.
          assert(U1 && U2);
          return Equal;
```
(the return Equal/Less/Greater would be replaced with `continue`/`return true`/`return false` in the current algorithm/directly inside the loop, but I'd moved the code into a lambda to make comparing the two comparison implementations easier)

This implementation doesn't have any cross-checking between different types of elements - clearly isolates and orders the different orderings making it easy to inspect/understand how the different groupings are implemented, I think?

**Comment by alexey-bataev** [comment]
Better to have Instructions < undefs< constants, better to combine insts/undefs first.
I would appreciate it if you can improve/fix it and prepare a patch.

**Comment by dwblaikie** [comment]
The behavior-preserving change is easy enough as outlined in my previous comment and I can send that for review, but changing the order as you requested does regress a couple of tests:

test/Transforms/SLPVectorizer/X86/reused-undefs.ll: https://www.diffchecker.com/V7BSldjt/
test/Transforms/SLPVectorizer/X86/phi-undef-input.ll: https://www.diffchecker.com/pctnAozF/

Second one looks like harmless reordering, but the first one looks ilke maybe a real regression? (or maybe it's an improvement? it's more lines, due to some more phi instructions at the end - but it did shrink the vectors from 8 to 4)?

In any case, changing the behavior here's a bit out of my depth/domain - mostly just wanted to clean up the code/make it easier to understand, more robust, etc. So perhaps I can send the behavior-preserving version, and if you want to do the reordering & handle the bugs/fixes/regressions/improvements in a follow up, that'd be OK?

**Comment by alexey-bataev** [comment]
Yes, appreciate your help!

**Comment by dwblaikie** [comment]
Sent #84966 which is the behavior preserving simplification/refactoring. (approved/committed as a843f26)

The reordering also makes the code a bit tidier (because the undef comes before constant, so there's no need to check "constant non-undef" explicitly - just instructions, then undef, then constants). I made a draft PR with the changes, linking back to the discussion here, commenting on the couple of failing tests, etc. If you want to copy that PR and address the issues, you're welcome to. #84969

**Comment by dwblaikie** [comment]
Going to close this issue, as I'm pretty confident with a843f26 this part of the comparator should be correct, at least.

**Comment by dwblaikie** [comment]
Hmm, seeing libc++ assertions in this comparator again (not sure if they are problems that survived the previous work, or a recent regression)... :/ Looking into it.
